### PR TITLE
*: add a flag EncodeType for response to use Arrow/Default format to decode.

### PR DESCRIPTION
**Pull request by wshwsh12**
Signed-off-by: wshwsh12 <793703860@qq.com>

###  What have you changed?

Please explain in detail what the changes are in this PR and why they are needed:

Now the `SelectResponse` use two field to save Arrow/Default format data. But they are both bytes array and can use the same field. We can add field `EncodeType` instead of field `RowBatchData`.
So remove the field `RowBatchData` and add the flag field `EncodeType`.

###  What is the type of the changes?

Pick one of the following and delete the others:

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Integration test

###  Refer to a related PR or issue link (optional)

TiDB: https://github.com/pingcap/tidb/pull/12536
TiPB: https://github.com/pingcap/tipb/pull/135

### CODE REVIEW

**Comment by wshwsh12** [comment]
I change the proto and adjust tidb and tikv's code.
`integration-compatibility-test` will use both tikv-master and tikv-pr/5584(this pr) to run test. And one of them will fail.
If the two related pr merge, the test will pass.

File: components/tidb_query/src/batch/runner.rs [hunk 1]
```diff
@@ -393,29 +393,29 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
                     // Although `schema()` can be deeply nested, it is ok since we process data in
                     // batch.
                     match self.encode_type {
-                        EncodeType::TypeDefault => {
-                            data.reserve(result.physical_columns.maximum_encoded_size(
+                        EncodeType::TypeArrow => {
```
Comment (breeswish) [inline components/tidb_query/src/batch/runner.rs]: How about:

```rust
let response_encode_type;
match self.desired_encode_type {
    EncodeType::TypeArrow => {
        response_encode_type = EncodeType::TypeArrow;
        ....
    }
    _ => {
        // For the default or unsupported encode type, use datum format.
        response_encode_type = EncodeType::TypeDefault;
    }
}
...
sel_resp.set_encode_type(response_encode_type);
```

We will only need to change code in one place when a new encoding is added.

**Comment by breeswish** [comment]
I think compatibility test will never pass, since we made a breaking change in the protocol itself. Let's ignore it?

File: components/tidb_query/src/batch/runner.rs (hunk 1 above)
Comment (wshwsh12) [reply components/tidb_query/src/batch/runner.rs]: Ok. I update the code. PTAL again.

**Comment by wshwsh12** [comment]
> I think compatibility test (against current master) will never pass, since we made a breaking change in the protocol itself. However it is compatible with 3.0 because the breaking change has not been made over 3.0. Let's ignore this failed test?

I also think so. Ignore this failed test is ok.

**Comment by breeswish** [comment]
@wshwsh12 Nice job. I have given my approval and you need another one to merge this PR.

**Comment by sticnarf** [comment]
By the way, is it a problem that tipb does not have a branch for version control? I worry the tipb version in a release branch may be upgraded accidentally.

**Comment by breeswish** [comment]
@sticnarf 3.0 branch TiKV should rely on 3.0 branch tipb.

**Comment by sticnarf** [comment]
> @sticnarf 3.0 branch TiKV should rely on 3.0 branch tipb.

Yes. That's what I hope it were. But actually it is not specified now: https://github.com/tikv/tikv/blob/release-3.0/Cargo.toml#L137
:(

**Comment by breeswish** [comment]
@sticnarf That's horrible!!! @BusyJay PTAL

**Comment by breeswish** [comment]
@wshwsh12 I believe you need to open a small PR that upgrades the tipb dependency in TiDB and merge that PR in order to make integration test not broken after this PR is merged, right?

**Pull request merged**